enterprise/server/util/cpuset/cpuset.go

```diff
@@ -75,60 +75,58 @@ func Format[I constraints.Integer](cpus ...I) string {
 	return strings.Join(cpuStrings, ",")
 }
 
-func parseListFormat(s string) ([]cpuInfo, error) {
-	// Example: "0-1,3" is parsed as []int{0, 1, 3}
-	var nodes []cpuInfo
-	var physicalID int
+// parseCPUs parses a string like "0,1-2" or "0:0-127,1:128-255" to cpuInfos.
+//
+// This string is a comma-separated list of "NODE:CPU_RANGE" pairs where the
+// "NODE:" prefix is optional.
+//
+// If the NODE isn't set, physicalIDs is used to map processor ids to physical
+// ids. Otherwise, the returned CPUs will have physicalID set to -1.
+func parseCPUs(s string, physicalIDs map[int]int) ([]cpuInfo, error) {
+	var cpus []cpuInfo
 	nodeRanges := strings.Split(s, ",")
 	for _, r := range nodeRanges {
-		startStr, endStr, _ := strings.Cut(r, "-")
-		if strings.Contains(startStr, ":") {
-			var numaStr string
-			numaStr, startStr, _ = strings.Cut(startStr, ":")
+		physicalID := -1
+		if numaStr, rangeStr, ok := strings.Cut(r, ":"); ok {
 			numaID, err := strconv.Atoi(numaStr)
 			if err != nil {
-				return nodes, fmt.Errorf("malformed file contents")
+				return cpus, fmt.Errorf("malformed NUMA node %q", numaStr)
 			}
 			physicalID = numaID
+			r = rangeStr
 		}
+		startStr, endStr, _ := strings.Cut(r, "-")
 		start, err := strconv.Atoi(startStr)
 		if err != nil {
-			return nodes, fmt.Errorf("malformed file contents")
+			return cpus, fmt.Errorf("malformed CPU start index %q", startStr)
 		}
 		end := start
 		if endStr != "" {
 			n, err := strconv.Atoi(endStr)
 			if err != nil {
-				return nodes, fmt.Errorf("malformed file contents")
+				return cpus, fmt.Errorf("malformed CPU end index %q", endStr)
 			}
 			if n < start {
-				return nodes, fmt.Errorf("malformed file contents")
+				return cpus, fmt.Errorf("invalid CPU range end index %d: must exceed start index %d", n, start)
 			}
 			end = n
 		}
-		for node := start; node <= end; node++ {
-			nodes = append(nodes, cpuInfo{
-				processor:  node,
+		for processor := start; processor <= end; processor++ {
+			info := cpuInfo{
+				processor:  processor,
 				physicalID: physicalID,
-			})
+			}
+			if info.physicalID == -1 {
```

Comment:
This might be overkill, but we could do this check even if the user did specify a node. I think this would've prevented that bug when Iain set an invalid value for the node?

Reply:
+1

Reply:
Done. The tests were specifying invalid nodes that don't necessarily exist on the machine being tested, so I had to update the NewLeaser to allow overriding the system CPUs, which resulted in a lot of misc diffs. I think everything is in good shape now though.